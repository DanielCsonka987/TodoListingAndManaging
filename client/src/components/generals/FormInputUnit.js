import React, { Component } from 'react'
import { Transition } from 'react-transition-group'

const outerWrapperStyle = {
    position:"relative",

}
const innerWrapperStyle = {
    padding: "0.2rem 1rem",

    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'space-between'
}
const inputNumberStyle = {
    width: '2.5rem'
}
const inputTextStyle = {
    width: '45%'
}
const bubbleStyle = {
    position: 'absolute',
    zIndex: '3',
    opacity: '0.0',
    maxWidth: '6.5rem',
    padding: '0.4rem',
    top: '-1.2rem',
    right: '-7rem',
    background: '#FFF0C3',
    borderRadius: '13%',
    border: '1px solid grey',
    fontSize: '0.8rem',
    transition: 'opacity 16000ms ease'
}

class FormInputUnit extends Component{
    constructor(props){
        super(props)
        this.showBubbleOnScreen = this.showBubbleOnScreen.bind(this)
        this.removeBubbleOnScreen = this.removeBubbleOnScreen.bind(this)
        this.state = {
            bubbleAppearance: false,
        }
    }
    showBubbleOnScreen(){
      this.setState({ bubbleAppearance: true })
    }
    removeBubbleOnScreen(){
      this.setState({ bubbleAppearance: false })
    }
    render(){
        const hiddenBubble = this.state.bubbleAppearance?
            this.props.children? <span key={1} style={ bubbleStyle } >{this.props.children}</span> 
                : <span key={1}></span>
            : 
            <span key={1}></span>
        let inputStylingConfig = {}
        if(this.props.type === 'number'){
            inputStylingConfig =  inputNumberStyle
        }else{
            inputStylingConfig.width = this.props.neededInputWidth || inputTextStyle.width
        }
        let defMinMax = {};
        if(this.props.numberMinMax){
            defMinMax.min = this.props.numberMinMax[0]
            defMinMax.max = this.props.numberMinMax[1]
        }
        
        return (
            <div style={ {...outerWrapperStyle, ...this.props.additWrapperStyles} } 
                className={this.props.classes}>
                <div style={innerWrapperStyle}>
                    <label className='dataLabelMarking' htmlFor={this.props.id}>{this.props.label}</label>
                    <input style={ inputStylingConfig }
                        name={this.props.name} type={this.props.type}
                        id={this.props.id} value={this.props.value} { ...defMinMax }
                        onChange={this.props.funcChange}  onKeyPress={this.props.funcHitEnter}
                        onFocus={this.showBubbleOnScreen} onMouseOver={this.showBubbleOnScreen}
                        onBlur={this.removeBubbleOnScreen } onMouseOut={this.removeBubbleOnScreen}
                    />
                </div>
                <Transition 
                    transitionentertimeout={500}
                    transitionleavetimeout={500}
                    timeout={500}
                >
                    { hiddenBubble  }
                </Transition >
            </div>
            
        )
    }
}

export default FormInputUnit;