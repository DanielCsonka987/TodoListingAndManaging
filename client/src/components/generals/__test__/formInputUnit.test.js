import React from 'react'
import { waitFor, fireEvent, screen, render } from '@testing-library/react'
import FormInputUnit from '../FormInputUnit'
import '@testing-library/jest-dom/extend-expect'
import renderer from 'react-test-renderer';
import { toMatchDiffSnapshot } from 'snapshot-diff'
import { expect as chaiExpect } from 'chai';
import userEvent from '@testing-library/user-event'

describe('Form input test', ()=>{
    test('Component set to textbox without popup bubble', async ()=>{
        const changeFunc = jest.fn()
        const enterFunc = jest.fn()
        const comp = <FormInputUnit
            label='Username area' classes='stgNoticable'
            type='text' name='username' id='unameLabelBound'
            value='stg' funcChange={changeFunc} funcHitEnter={enterFunc}
        >
        </FormInputUnit>

        const tree = renderer.create(comp).toJSON()
        expect(tree).toMatchSnapshot()

        const { container, unmount, debug } = render(comp)

        const itemOuterContainer = container.firstChild
        expect(itemOuterContainer).toHaveClass('stgNoticable')

        const theInputItem1 = screen.getByDisplayValue('stg')
        const theInputItem2 = screen.getByLabelText('Username area')
        expect(theInputItem1).toBe(theInputItem2)
        expect(theInputItem1).toHaveAttribute('name', 'username')
        expect(theInputItem1).toHaveAttribute('type', 'text')
        expect(theInputItem1).toHaveAttribute('id', 'unameLabelBound')

        const emptyBubble = itemOuterContainer.lastChild
        expect(emptyBubble).toHaveTextContent('')
        
        fireEvent.change(theInputItem2, { target: { value: 'different' } })
        await waitFor(()=>{
            expect(changeFunc.mock.calls.length).toBe(1)
        })

        /*fireEvent.keyPress(theInputItem1, { key: 'Enter', code: 'Enter' })
        await waitFor(()=>{
            expect(enterFunc.mock.calls.length).toBe(1)
        })*/
        userEvent.type(theInputItem1, 'another')
        expect(enterFunc.mock.calls.length).toBe(7)

        unmount()

    })

    test('Component set to textbox with little popup bubble', async ()=>{
        const changeFunc = jest.fn()
        const enterFunc = jest.fn()
        const bubbleText = 'Test text to test this'
        const comp = <FormInputUnit
            label='Username area' classes='stgNoticable'
            type='textbox' name='username' id='unameLabelBound'
            value='stg' funcChange={changeFunc} funcHitEnter={enterFunc}
        >
            {bubbleText}
        </FormInputUnit>

        const tree = renderer.create(comp).toJSON()
        expect(tree).toMatchSnapshot()

        const { container, unmount, debug } = render(comp)

        const itemOuterContainer = container.firstChild
        expect(itemOuterContainer).toHaveClass('stgNoticable')
        expect(screen.queryByText(bubbleText)).not.toBeInTheDocument()

        const theInputItem2 = screen.getByLabelText('Username area')
        fireEvent.focus(theInputItem2)
        await waitFor(()=>{
            const bubbleTag = screen.queryByText(bubbleText)
            expect(bubbleTag).toBeInTheDocument()
            expect(bubbleTag).toHaveStyle('position: absolute')
        })
        fireEvent.focusOut(theInputItem2)
        await waitFor(()=>{
            const bubbleTag = screen.queryByText(bubbleText)
            expect(bubbleTag).not.toBeInTheDocument()
        })
        unmount()
    })
})