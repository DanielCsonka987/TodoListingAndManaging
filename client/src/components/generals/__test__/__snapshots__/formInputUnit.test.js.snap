// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form input test Component set to textbox with little popup bubble 1`] = `
<div
  className="stgNoticable"
  style={
    Object {
      "position": "relative",
      "width": "100%",
    }
  }
>
  <div
    style={
      Object {
        "display": "flex",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "padding": "0.2rem 1rem",
      }
    }
  >
    <label
      className="dataLabelMarking"
      htmlFor="unameLabelBound"
    >
      Username area
    </label>
    <input
      id="unameLabelBound"
      name="username"
      onBlur={[Function]}
      onChange={[MockFunction]}
      onFocus={[Function]}
      onKeyPress={[MockFunction]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "width": "45%",
        }
      }
      type="textbox"
      value="stg"
    />
  </div>
  <span
    transitionentertimeout={500}
    transitionleavetimeout={500}
  />
</div>
`;

exports[`Form input test Component set to textbox without popup bubble 1`] = `
<div
  className="stgNoticable"
  style={
    Object {
      "position": "relative",
      "width": "100%",
    }
  }
>
  <div
    style={
      Object {
        "display": "flex",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "padding": "0.2rem 1rem",
      }
    }
  >
    <label
      className="dataLabelMarking"
      htmlFor="unameLabelBound"
    >
      Username area
    </label>
    <input
      id="unameLabelBound"
      name="username"
      onBlur={[Function]}
      onChange={[MockFunction]}
      onFocus={[Function]}
      onKeyPress={[MockFunction]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "width": "45%",
        }
      }
      type="text"
      value="stg"
    />
  </div>
  <span
    transitionentertimeout={500}
    transitionleavetimeout={500}
  />
</div>
`;
